src/socket/udp_socket.cr

```diff
@@ -88,4 +88,139 @@ class UDPSocket < IPSocket
+      level = LibC::IPPROTO_IP
+    when Family::INET6
+      optname = LibC::IPV6_MULTICAST_HOPS
+      level = LibC::IPPROTO_IPV6
+    else
+      raise "Unsupported IP address family: #{family}"
+    end
+    getsockopt optname, 0, level
+  end
+
+  # The multicast hops option controls the `hoplimit` field on uni-cast packets.
+  # If `-1` is specified, the kernel will use a default value.
+  # If a value of `0` to `255` is specified, the packet will have the specified
+  # value as `hoplimit`. Other values are considered invalid.
+  # Datagrams with a `hoplimit` of `1` are not forwarded beyond the local network.
+  # Multicast datagrams with a `hoplimit` of `0` will not be transmitted on any
+  # network, but may be delivered locally if the sending host belongs to the
+  # destination group and multicast loopback is enabled.
+  def multicast_hops=(val : Int)
+    case @family
+    when Family::INET
+      optname = LibC::IP_MULTICAST_TTL
+      level = LibC::IPPROTO_IP
+    when Family::INET6
+      optname = LibC::IPV6_MULTICAST_HOPS
+      level = LibC::IPPROTO_IPV6
+    else
+      raise "Unsupported IP address family: #{family}"
+    end
+    setsockopt optname, val, level
+    val
+  end
+
+  # For hosts with multiple interfaces, each multicast transmission is sent
+  # from the primary network interface. This function overrides the default
+  # IPv4 interface address for subsequent transmissions. Setting the interface
+  # to `0.0.0.0` will select the default interface.
+  def multicast_interface(address : IPAddress)
+    if @family == Family::INET
+      setsockopt LibC::IP_MULTICAST_IF, address.addr4.not_nil!, LibC::IPPROTO_IP
+    else
+      raise "For use with IPv4 only. Unsupported IP address family: #{family}"
+    end
+  end
+
+  # For hosts with multiple interfaces, each multicast transmission is sent
+  # from the primary network interface. This function overrides the default
+  # IPv6 interface for subsequent transmissions. Setting the interface to
+  # index `0` will select the default interface.
+  def multicast_interface(index : UInt32)
+    if @family == Family::INET6
+      setsockopt LibC::IPV6_MULTICAST_IF, index, LibC::IPPROTO_IPV6
+    else
+      raise "For use with IPv6 only. Unsupported IP address family: #{family}"
+    end
+  end
+
+  # A host must become a member of a multicast group before it can receive
+  # datagrams sent to the group.
+  def join_group(ip : IPAddress)
+    group_modify(ip, true)
+  end
+
+  # Drops membership to the specified group. Memberships are automatically
+  # dropped when the socket is closed or the process exits.
+  def leave_group(ip : IPAddress)
+    group_modify(ip, false)
+  end
+
+  private def group_modify(ip, operation)
+    resp = case ip.family
+           when Family::INET6
+             operation = operation ? LibC::IPV6_JOIN_GROUP : LibC::IPV6_LEAVE_GROUP
+             req = LibC::Ipv6Mreq.new
+             req.ipv6mr_multiaddr = ip.addr6.not_nil!
```

Comment:
This nil check should be guarded to provide a specific error when a non-matching address family is used. Ditto below.